### ISSUE TITLE
ability to don't move imports separated by "actual" code

### ISSUE THREAD

**Issue by juliusmarminge** [IanVS/prettier-plugin-sort-imports#70]
Would it be possible to support something like this:

```ts
import x from "x";
import y from "y";

function z() {
  // ...
}

import foo from "foo"; // <-- don't move this import to the top
import bar from "bar"; //     but it's fine sorting foo, bar
function baz() {
 // ...
}
```

i.e. if i separate my file into separate "chunks" - i don't want all the imports at the top, but they can still be sorted in each chunk. so here x, y would be sorted and foo, bar would be sorted - but foo, bar shouldn't be hoisted to the top.

makes sense?

**Comment by IanVS** [comment]
Hmmm, that's interesting.  Can you talk more about the use case here?  For example, `function z` would normally be sorted below all of the imports to avoid confusion, since all of the imports will be "hoisted" above the function declaration.  This is why static imports are normally placed at the top of the file, to match the evaluation order of the runtime.

**Comment by juliusmarminge** [comment]
We have this in create-t3-app: https://github.com/t3-oss/create-t3-app/blob/next/cli/template/extras/src/server/api/trpc/with-auth-prisma.ts

That file is split into 3 sections and each section has their "own imports". Ofc not the end of the world to hoist all imports to the top but I'd still like the option to not hoist them

**Comment by IanVS** [comment]
And yeah, the docblocks travel with the import as demonstrated in that diff.  That's intentional, since it's assumed that a leading comment like that is associated with the import it's directly above.  But I understand it's not what you want in this case.

**Comment by fbartho** [comment]
Having written a lot of Javascript before the modern era of official tooling, I recognize this pattern as something that has become pretty strongly assigned to some other tool in your stack. This is basically something intended for your compiler or bundler or your runtime VM.

Bundlers often have a setting where imports are converted from a global-scope entry at the top of file to a once-only asynchronously executed `require` inlined in the file to every point of use. (This setting is useful for only loading the code necessary to launch your app, and letting other modules execute later).

Some runtimes or module systems will crash if you try to import a circular dependency, while other systems are fine with circular dependencies -- as long as you don't try to interact with the exported attribute until it's fully loaded. Tree shaking tools also interact with this.

Having our plugin encourage intentionally interlacing function-definitions (global-scope code) with global-scope import statements encourages developer proximity with these edge cases.

Further, I think this is not really recommended JS practice in general, so this would have to be a configurable option defaulted-to-disabled.

**In conclusion**: I believe this requested feature could accidentally expose differences between ES-Modules and CommonJS, or even between different runtimes, which… would not be ideal -- leading to significant confusion, and difficult Developer Debugging & complex configuration.

I don't think this feature should be in this plugin, I'm sorry @juliusmarminge.

**Comment by IanVS** [comment]
I agree with @fbartho that we don't want to explicitly add this as an option.  However, if there's a small tweak we can make or a new type of ignore comment we can add that would solve your use case, I'm willing to explore that.

**Comment by fbartho** [comment]
@juliusmarminge I'm going to close this ticket as `"Not Planned"` for now, but if you or other future Github Users want to revisit this discussion, we can reopen it if you have new perspectives to share!
